sharktank/sharktank/types/ocp_floats.py

```diff
@@ -299,3 +297,94 @@ def float32_to_fp4_e2m1(values: torch.Tensor) -> torch.Tensor:
     fp4_indices = torch.argmin(abs_diff, dim=-1)
 
     return fp4_indices.to(torch.uint8)
+
+
+from sharktank.kernels.base import *
+from sharktank.kernels.mlir_kernel import *
+
+N = DynDim.N
+M32 = StaticDim.M32(32)
+M16 = StaticDim.M16(16)
+
+I_DTYPE = Dtype.I_DTYPE
+U8_DTYPE = Dtype.U8(torch.uint8)
+F32_DTYPE = Dtype.F32(torch.float32)
+
+
+@mlir_kernel(
+    inputs=(MLIRTensor[N, M32, I_DTYPE],),
+    results=(MLIRTensor[N, U8_DTYPE], MLIRTensor[N, M16, U8_DTYPE]),
+)
+def dynamic_quantize_to_fp4(values, result_scales=None, result_quantized=None):
+    mlir = """
+    module {
+    util.func private @{{kernel_name}}(%values : !values) -> (!result_scales, !result_quantized) {
+      %c0 = arith.constant 0 : index
+      %c1 = arith.constant 1 : index
+      %c4_f32 = arith.constant 4.0 : f32
+      %neg_inf = arith.constant -3.402823e+38 : f32
+
+      %batch_dim = tensor.dim %values, %c0 : !values
+      %empty_scales = tensor.empty(%batch_dim) : !result_scales
+      %empty_max = tensor.empty(%batch_dim) : tensor<?xf32>
+      %empty_quantized = tensor.empty(%batch_dim) : !result_quantized
+
+      // Step 1: Find max absolute value per block (reduction)
+      %init_max = linalg.fill ins(%neg_inf : f32) outs(%empty_max : tensor<?xf32>) -> tensor<?xf32>
+      %block_max = linalg.reduce ins(%values : !values) outs(%init_max : tensor<?xf32>) dimensions = [1]
+        (%in: f32, %init: f32) {
+          %abs_val = math.absf %in : f32
+          %max_val = arith.maximumf %abs_val, %init : f32
+          linalg.yield %max_val : f32
+        }
+
+      // Step 2: Convert max to FE8M0 scales
+      %scales = linalg.generic {
+        indexing_maps = [
+          affine_map<(d0) -> (d0)>,
+          affine_map<(d0) -> (d0)>
+        ],
+        iterator_types = ["parallel"]
+      } ins(%block_max : tensor<?xf32>) outs(%empty_scales : !result_scales) {
+      ^bb0(%max_val: f32, %out_scale: i8):
+        %biased_scale = arith.divf %max_val, %c4_f32 : f32
+        %fe8m0_scale = arith.truncf %biased_scale : f32 to f8E8M0FNU
+        %scale_i8 = arith.bitcast %fe8m0_scale : f8E8M0FNU to i8
+        linalg.yield %scale_i8 : i8
+      } -> !result_scales
+
+      // Step 3: Quantize and pack FP4 values directly to bytes
+      %empty_packed = tensor.empty(%batch_dim) : !result_quantized
+      %quantized = linalg.generic {
+        indexing_maps = [
+          affine_map<(d0, d1) -> (d0, d1 * 2)>,
+          affine_map<(d0, d1) -> (d0, d1 * 2 + 1)>,
+          affine_map<(d0, d1) -> (d0)>,
+          affine_map<(d0, d1) -> (d0, d1)>
+        ],
+        iterator_types = ["parallel", "parallel"]
+      } ins(%values, %values, %scales : !values, !values, !result_scales) outs(%empty_packed : !result_quantized) {
+      ^bb0(%val_low: f32, %val_high: f32, %scale: i8, %out: i8):
```

Comment:
This won't pattern-match right for the AMDGPU backend

To use device-side instructions we need:

1. A _vector_ scaling_truncf - that is, taking a `vector<2xf32>`
2. The scale argument should be a broadcast vector (so a `vector<2xf8E8M0FNU>`)
3. ... except that the hardware intrinsics actually take a `f32` value and do the conversion to f8E8M0FNU implicitly, so you don't need all the casting. (Except that the MFMAs do need the casting)

Reply:
Fixed after extensive discussion elsewhere.